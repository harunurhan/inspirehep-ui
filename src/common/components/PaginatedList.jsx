import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { List, Pagination } from 'antd';
import Immutable from 'immutable';

class PaginatedList extends Component {
  static getDerivedStateFromProps(nextProps, prevState) {
    const { items, pageSize } = nextProps;
    const page = 1;
    const pageItems = PaginatedList.getPageItems(items, page, pageSize);
    return {
      ...prevState,
      pageItems,
      page,
      total: items.size,
    };
  }

  static getPageItems(items, page, pageSize) {
    const endIndex = page * pageSize;
    const startIndex = endIndex - pageSize;
    return items.slice(startIndex, endIndex);
  }

  static getPaginationRangeInfo(total, range) {
    return `${range[0]}-${range[1]} of ${total}`;
  }

  constructor(props) {
    super(props);
    this.state = {};

    this.onPageChange = this.onPageChange.bind(this);
  }

  onPageChange(page) {
    const { pageSize, items } = this.props;
    const pageItems = PaginatedList.getPageItems(items, page, pageSize);
    this.setState({
      page,
      pageItems,
    });
  }

  renderPagination() {
    const { pageSize, loading } = this.props;
    const { total, page } = this.state;
    return (
      <Pagination
        current={page}
        onChange={this.onPageChange}
        total={total}
        pageSize={pageSize}
        loading={loading}
        showTotal={PaginatedList.getPaginationRangeInfo}
      />
    );
  }

  render() {
    const { renderItem, title, items } = this.props;
    const { pageItems } = this.state;
    return (
      items.size > 0 && (
        <List header={title} footer={this.renderPagination()}>
          {pageItems.map(item => renderItem(item))}
        </List>
      )
    );
  }
}

PaginatedList.propTypes = {
  title: PropTypes.node,
  items: PropTypes.instanceOf(Immutable.List),
  renderItem: PropTypes.func.isRequired,
  loading: PropTypes.bool,
  pageSize: PropTypes.number,
};

PaginatedList.defaultProps = {
  items: Immutable.List(),
  title: null,
  pageSize: 25,
  loading: false,
};

export default PaginatedList;
