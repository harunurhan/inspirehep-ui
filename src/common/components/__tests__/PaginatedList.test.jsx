import React from 'react';
import { shallow } from 'enzyme';
import { fromJS, Range } from 'immutable';
import { List } from 'antd';

import PaginatedList from '../PaginatedList';

describe('PaginatedList', () => {
  it('renders first page initially with default pageSize', () => {
    const items = Range(1, 100).toList();
    const wrapper = shallow(
      <PaginatedList
        items={items}
        renderItem={item => <List.Item key={item}>{item}</List.Item>}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders first page initially with pageSize', () => {
    const items = Range(1, 100).toList();
    const wrapper = shallow(
      <PaginatedList
        items={items}
        renderItem={item => <List.Item key={item}>{item}</List.Item>}
        pageSize={10}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the new page on page change', () => {
    const items = Range(1, 100).toList();
    const page = 2;
    const wrapper = shallow(
      <PaginatedList
        items={items}
        renderItem={item => <List.Item key={item}>{item}</List.Item>}
        pageSize={10}
      />
    );
    wrapper.instance().onPageChange(page);
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });

  it('does not render at all if empty', () => {
    const items = fromJS([]);
    const wrapper = shallow(
      <PaginatedList
        items={items}
        renderItem={item => <List.Item key={item}>{item}</List.Item>}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  describe('getPageItems', () => {
    it('returns items for a page', () => {
      const items = fromJS([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
      const page = 2;
      const pageSize = 3;
      const expected = fromJS([4, 5, 6]);
      const result = PaginatedList.getPageItems(items, page, pageSize);
      expect(result).toEqual(expected);
    });

    it('returns items for the first page', () => {
      const items = fromJS([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
      const page = 1;
      const pageSize = 3;
      const expected = fromJS([1, 2, 3]);
      const result = PaginatedList.getPageItems(items, page, pageSize);
      expect(result).toEqual(expected);
    });

    it('returns only remaining items for the last page', () => {
      const items = fromJS([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
      const page = 4;
      const pageSize = 3;
      const expected = fromJS([10]);
      const result = PaginatedList.getPageItems(items, page, pageSize);
      expect(result).toEqual(expected);
    });
  });
});
