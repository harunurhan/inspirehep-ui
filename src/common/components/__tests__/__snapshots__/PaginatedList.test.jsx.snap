// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginatedList does not render at all if empty 1`] = `""`;

exports[`PaginatedList renders as loading if set 1`] = `
<List
  bordered={false}
  dataSource={Array []}
  footer={
    <Pagination
      current={1}
      loading={true}
      onChange={[Function]}
      pageSize={10}
      prefixCls="ant-pagination"
      selectPrefixCls="ant-select"
      showTotal={[Function]}
      total={99}
    />
  }
  header={null}
  loading={false}
  pagination={false}
  prefixCls="ant-list"
  split={true}
>
  <Item
    key="1"
  >
    1
  </Item>
  <Item
    key="2"
  >
    2
  </Item>
  <Item
    key="3"
  >
    3
  </Item>
  <Item
    key="4"
  >
    4
  </Item>
  <Item
    key="5"
  >
    5
  </Item>
  <Item
    key="6"
  >
    6
  </Item>
  <Item
    key="7"
  >
    7
  </Item>
  <Item
    key="8"
  >
    8
  </Item>
  <Item
    key="9"
  >
    9
  </Item>
  <Item
    key="10"
  >
    10
  </Item>
</List>
`;

exports[`PaginatedList renders first page initially with default pageSize 1`] = `
<List
  bordered={false}
  dataSource={Array []}
  footer={
    <Pagination
      current={1}
      loading={false}
      onChange={[Function]}
      pageSize={25}
      prefixCls="ant-pagination"
      selectPrefixCls="ant-select"
      showTotal={[Function]}
      total={99}
    />
  }
  header={null}
  loading={false}
  pagination={false}
  prefixCls="ant-list"
  split={true}
>
  <Item
    key="1"
  >
    1
  </Item>
  <Item
    key="2"
  >
    2
  </Item>
  <Item
    key="3"
  >
    3
  </Item>
  <Item
    key="4"
  >
    4
  </Item>
  <Item
    key="5"
  >
    5
  </Item>
  <Item
    key="6"
  >
    6
  </Item>
  <Item
    key="7"
  >
    7
  </Item>
  <Item
    key="8"
  >
    8
  </Item>
  <Item
    key="9"
  >
    9
  </Item>
  <Item
    key="10"
  >
    10
  </Item>
  <Item
    key="11"
  >
    11
  </Item>
  <Item
    key="12"
  >
    12
  </Item>
  <Item
    key="13"
  >
    13
  </Item>
  <Item
    key="14"
  >
    14
  </Item>
  <Item
    key="15"
  >
    15
  </Item>
  <Item
    key="16"
  >
    16
  </Item>
  <Item
    key="17"
  >
    17
  </Item>
  <Item
    key="18"
  >
    18
  </Item>
  <Item
    key="19"
  >
    19
  </Item>
  <Item
    key="20"
  >
    20
  </Item>
  <Item
    key="21"
  >
    21
  </Item>
  <Item
    key="22"
  >
    22
  </Item>
  <Item
    key="23"
  >
    23
  </Item>
  <Item
    key="24"
  >
    24
  </Item>
  <Item
    key="25"
  >
    25
  </Item>
</List>
`;

exports[`PaginatedList renders first page initially with pageSize 1`] = `
<List
  bordered={false}
  dataSource={Array []}
  footer={
    <Pagination
      current={1}
      loading={false}
      onChange={[Function]}
      pageSize={10}
      prefixCls="ant-pagination"
      selectPrefixCls="ant-select"
      showTotal={[Function]}
      total={99}
    />
  }
  header={null}
  loading={false}
  pagination={false}
  prefixCls="ant-list"
  split={true}
>
  <Item
    key="1"
  >
    1
  </Item>
  <Item
    key="2"
  >
    2
  </Item>
  <Item
    key="3"
  >
    3
  </Item>
  <Item
    key="4"
  >
    4
  </Item>
  <Item
    key="5"
  >
    5
  </Item>
  <Item
    key="6"
  >
    6
  </Item>
  <Item
    key="7"
  >
    7
  </Item>
  <Item
    key="8"
  >
    8
  </Item>
  <Item
    key="9"
  >
    9
  </Item>
  <Item
    key="10"
  >
    10
  </Item>
</List>
`;

exports[`PaginatedList renders the new page on page change 1`] = `
<List
  bordered={false}
  dataSource={Array []}
  footer={
    <Pagination
      current={2}
      loading={false}
      onChange={[Function]}
      pageSize={10}
      prefixCls="ant-pagination"
      selectPrefixCls="ant-select"
      showTotal={[Function]}
      total={99}
    />
  }
  header={null}
  loading={false}
  pagination={false}
  prefixCls="ant-list"
  split={true}
>
  <Item
    key="11"
  >
    11
  </Item>
  <Item
    key="12"
  >
    12
  </Item>
  <Item
    key="13"
  >
    13
  </Item>
  <Item
    key="14"
  >
    14
  </Item>
  <Item
    key="15"
  >
    15
  </Item>
  <Item
    key="16"
  >
    16
  </Item>
  <Item
    key="17"
  >
    17
  </Item>
  <Item
    key="18"
  >
    18
  </Item>
  <Item
    key="19"
  >
    19
  </Item>
  <Item
    key="20"
  >
    20
  </Item>
</List>
`;
